content/courses/Ada_For_The_C_Embedded_Developer/chapters/06_Translation.rst

```diff
@@ -725,9 +725,467 @@ above could also be literally translated to:
+complex data structures as a bitstream. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Serialization_ada
+
+    package Serializer is
+
+       type Bit_Field is array (Natural range <>) of Boolean with Pack;
+
+       procedure Transmit (B : Bit_Field);
+
+    end Serializer;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Serializer is
+
+       procedure Transmit (B : Bit_Field) is
+
+          procedure Show_Bit (V : Boolean) is
+          begin
+             case V is
+                when False => Put ("0");
+                when True  => Put ("1");
+             end case;
+          end Show_Bit;
+
+       begin
+          Put ("Bits: ");
+          for I in B'Range loop
+             Show_Bit (B (I));
+          end loop;
+          New_Line;
+       end Transmit;
+
+    end Serializer;
+
+    package My_Recs is
+
+       type Rec is record
+          V : Integer;
+          S : String (1 .. 3);
+       end record;
+
+    end My_Recs;
+
+    with Serializer;  use Serializer;
+    with My_Recs;     use My_Recs;
+
+    procedure Main is
+       R : Rec := (5, "abc");
+       B : Bit_Field (0 .. R'Size - 1) with Address => R'Address;
+    begin
+       Transmit (B);
+    end Main;
+
+In this example, the :ada:`Transmit` procedure from :ada:`Serializer` package
+displays the individual bits of a bit-field. We could have used this strategy
+to actually transmit the information as a bitstream. In the main application,
+we call :ada:`Transmit` for the object :ada:`R` of record type :ada:`Rec`.
+Since :ada:`Transmit` has the bit-field type as a parameter, we can use it
+for any type, as long as we have a corresponding bit-field representation.
+
+In C, we interpret the input pointer as an array of bytes, and then use
+shifting and masking to access the bits of that byte. Here, we use the
+:c:`char` type because it has a size of one byte in most platforms.
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Serialization_C
+
+    !my_recs.h
+    typedef struct {
+        int v;
+        char s[3];
+    } rec;
+
+    !serializer.h
+    void transmit (void *bits, int len);
+
+    !serializer.c
+    #include "serializer.h"
+
+    #include <stdio.h>
+    #include <assert.h>
+
+    void transmit (void *bits, int len)
+    {
+        int i, j;
+        char *c = (char *)bits;
+
+        assert(sizeof(char) == 1);
+
+        printf("Bits: ");
+        for (i = 0; i < len / (sizeof(char) * 8); i++)
+        {
+            for (j = 0; j < sizeof(char) * 8; j++)
+            {
+                printf("%d", c[i] >> j & 1);
+            }
+        }
+        printf("\n");
+    }
+
+    !bitfield_serialization.c
+    #include <stdio.h>
+
+    #include "my_recs.h"
+    #include "serializer.h"
+
+    int main(int argc, const char * argv[])
+    {
+        rec r = {5, "abc"};
```

Comment:
Should't the C string to be 4 characters len, to account for the \0 at the end? Otherwise it seems like we're writing one character to far...

Reply:
You're right. I wanted to keep the C version as close as possible to the Ada version, but the way I'm initializing the string for `r` is still wrong (`strncpy` could be used instead to copy only 3 characters). In any case, a non null-terminated array of characters is not a valid C string, so I'll correct the example.